x-pack/plugins/apm/public/components/shared/Stacktrace/index.tsx

```diff
@@ -4,54 +4,79 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
+import { EuiLink, EuiTitle } from '@elastic/eui';
+import { get, isEmpty } from 'lodash';
 import React, { PureComponent } from 'react';
 import styled from 'styled-components';
-import { isEmpty, get } from 'lodash';
-import CodePreview from '../../shared/CodePreview';
-import { Ellipsis } from '../../shared/Icons';
-import { units, px } from '../../../style/variables';
+import { Stackframe } from '../../../../typings/APMDoc';
+import { px, units } from '../../../style/variables';
+import { CodePreview } from '../../shared/CodePreview';
 import { EmptyMessage } from '../../shared/EmptyMessage';
-import { EuiLink, EuiTitle } from '@elastic/eui';
+// @ts-ignore
+import { Ellipsis } from '../../shared/Icons';
 import { FrameHeading } from './FrameHeading';
 
+export interface StackframeCollapsed extends Stackframe {
+  libraryFrame?: boolean;
+  stackframes?: Stackframe[];
+}
+
 const LibraryFrameToggle = styled.div`
   margin: 0 0 ${px(units.plus)} 0;
   user-select: none;
 `;
 
-const LibraryFrames = styled.div``;
-
-function getCollapsedLibraryFrames(stackframes) {
-  return stackframes.reduce((acc, stackframe) => {
+function getCollapsedLibraryFrames(
+  stackframes: Stackframe[]
+): StackframeCollapsed[] {
+  return stackframes.reduce((acc: any, stackframe: StackframeCollapsed) => {
     if (!stackframe.libraryFrame) {
       return [...acc, stackframe];
     }
 
     // current stackframe is library frame
-    const prevItem = acc[acc.length - 1];
+    const prevItem: StackframeCollapsed = acc[acc.length - 1];
     if (!get(prevItem, 'libraryFrame')) {
       return [...acc, { libraryFrame: true, stackframes: [stackframe] }];
     }
 
     return [
       ...acc.slice(0, -1),
-      { ...prevItem, stackframes: [...prevItem.stackframes, stackframe] }
+      {
+        ...prevItem,
+        stackframes: prevItem.stackframes
+          ? [...prevItem.stackframes, stackframe]
+          : [stackframe]
+      }
```

Comment:
Even though I originally wrote `getCollapsedLibraryFrames` I barely understand it now. Not because of your changes. It was just never very readable. So if you feel like simplifing/refactor/change it you are very welcome. And perhaps add a test for it 🙏

Reply:
i pulled it into it's own file and added a very basic test...next, i'll amend this PR and take a try at simplifying this function

Reply:
Sounds good 👍 
Coincidentally a [new issue was created](https://github.com/elastic/kibana/issues/26525) directly related to this. Probably makes sense to refactor and fix the issue at the same time.